frontend/src/app/data-entry/shared/episode-link-form/episode-link-form.component.ts

```diff
@@ -0,0 +1,174 @@
+    @Input({ required: true }) entityID!: string;
+    @Input({ required: true }) episodeID!: string;
+
+    /** Type of entity (agent, letter, etc.) */
+    @Input({ required: true }) entityType!: Entity;
+
+    /** Describes the direction of the link in the document.
+     *
+     * The header will show the model that is being linked *to*; the model being linked
+     * *from* should be clear from the context.
+     */
+    @Input() linkTo: LinkTo = 'episode';
+
+
+    data$: Observable<DataEntryEpisodeEntityLinkQuery | undefined>;
+    form = new FormGroup({
+        sourceMention: new FormControl<SourceMention>(SourceMention.Direct),
+        note: new FormControl<string>('', { nonNullable: true }),
+    });
+
+    sourceMentionOptions: SelectOptions<SourceMention> = sourceMentionSelectOptions();
+
+    private link$ = new Subject<UpdateEpisodeEntityLinkInput>();
+    private status$ = formStatusSubject();
+    private id = `episode-link-${nextID++}`;
+
+    actionIcons = actionIcons;
+
+    constructor(
+        query: DataEntryEpisodeEntityLinkGQL,
+        private updateMutation: DataEntryUpdateEpisodeEntityLinkGQL,
+        private formService: FormService,
+    ) {
+        this.formService.attachForm(this.id, this.status$);
+
+        this.data$ = this.link$.pipe(
+            switchMap((link) => query.watch(link).valueChanges),
+            map(result => result.data),
+            shareReplay(1),
+            takeUntilDestroyed(),
+        );
+        this.data$.subscribe(this.updateFormValues.bind(this));
+
+        this.form.valueChanges.pipe(
+            debounceTime(500),
+            distinctUntilChanged(_.isEqual),
+            skip(1),
+            tap(() => this.status$.next('loading')),
+            withLatestFrom(this.link$),
+            map(this.toMutationInput),
+            switchMap(this.makeMutation.bind(this)),
+            takeUntilDestroyed(),
+        ).subscribe(this.onMutationResult.bind(this));
+    }
+
+    get sourceMentionInputID() {
+        return this.id + '-source-mention-input';
+    }
+
+    get notesInputID() {
+        return this.id + '-notes-input';
+    }
+
+    get entityName(): string {
+        return entityTypeNames[this.entityType];
+    }
+
+    get linkedObjectName(): string {
+        return this.linkTo == 'episode' ? 'episode' : this.entityName;
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        if (changes['entityID'] || changes['episodeID'] || changes['entityType']) {
+            this.link$.next({
+                entity: this.entityID,
+                episode: this.episodeID,
+                entityType: this.entityType,
+            });
+        }
+    }
+
+    ngOnDestroy(): void {
+        this.link$.complete();
+        this.status$.complete();
+        this.formService.detachForm(this.id);
+    }
+
+    formUrl(linkTo: LinkTo, entityType: Entity): string {
+        if (linkTo == 'episode') {
+            return 'episodes';
+        }
+        const routes = {
```

Comment:
This object is very similar to `entityTypeNames` in `utils.ts`. For consistency's sake, please put this object there as well. (You can maybe merge them as well, or derive one from the other). If you've done that, you can replace this function's body with `return linkTo == 'episode' ? 'episodes' : entityTypeRoutes[entityType];`